pkg/uninstall/uninstall.go

```diff
@@ -0,0 +1,166 @@
+	Close()
+}
+
+type GeminiUninstaller struct {
+	// ip -> remotes
+	remotes map[string]*config.RemoteHost
+	// ip -> ssh clients
+	sshClients map[string]*ssh.Client
+
+	configurator config.Configurator // conf reader
+	executor     exec.Executor       // execute commands on remote host
+	upDataPath   map[string]string   // ip->up path
+
+	wg sync.WaitGroup
+
+	clusterOptions install.ClusterOptions
+}
+
+func NewGeminiUninstaller(ops install.ClusterOptions) Uninstall {
+	new := &GeminiUninstaller{
+		remotes:        make(map[string]*config.RemoteHost),
+		sshClients:     make(map[string]*ssh.Client),
+		configurator:   config.NewGeminiConfigurator(ops.YamlPath, "", ""),
+		upDataPath:     make(map[string]string),
+		clusterOptions: ops,
+	}
+	return new
+}
+
+func (s *GeminiUninstaller) Prepare() error {
+	var err error
+	if err = s.configurator.BuildConfig(); err != nil {
+		return err
+	}
+	conf := s.configurator.GetConfig()
+
+	if err = s.prepareRemotes(conf); err != nil {
+		fmt.Printf("Failed to establish SSH connections with all remote servers. The specific error is: %s\n", err)
+		return err
+	}
+	fmt.Println("Success to establish SSH connections with all remote servers.")
+
+	s.executor = exec.NewGeminiExecutor(s.sshClients)
+
+	return nil
+}
+
+func (s *GeminiUninstaller) prepareRemotes(c *config.Config) error {
+	if c == nil {
+		return util.ErrUnexpectedNil
+	}
+
+	for ip, ssh := range c.SSHConfig {
+		s.remotes[ip] = &config.RemoteHost{
+			Ip:       ip,
+			SSHPort:  ssh.Port,
+			User:     s.clusterOptions.User,
+			Password: s.clusterOptions.Password,
+			KeyPath:  s.clusterOptions.Key,
+			Typ:      s.clusterOptions.SshType,
+		}
+
+		s.upDataPath[ip] = ssh.UpDataPath
+	}
+
+	if err := s.tryConnect(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *GeminiUninstaller) tryConnect() error {
+	for ip, r := range s.remotes {
+		var err error
+		var sshClient *ssh.Client
+		switch r.Typ {
+		case config.SSH_PW:
+			sshClient, err = util.NewSSH_PW(r.User, r.Password, r.Ip, r.SSHPort)
+		case config.SSH_KEY:
+			sshClient, err = util.NewSSH_Key(r.User, r.KeyPath, r.Ip, r.SSHPort)
+
+		}
+		if err != nil {
+			return err
+		}
+		s.sshClients[ip] = sshClient
+	}
+	return nil
+}
+
+func (s *GeminiUninstaller) Run() error {
+	if s.executor == nil {
+		return util.ErrUnexpectedNil
+	}
+
+	errChan := make(chan error, len(s.remotes))
+	var wgp sync.WaitGroup
+	wgp.Add(2)
+
+	go func() {
+		defer wgp.Done()
+		s.wg.Add(len(s.remotes))
+		for ip := range s.remotes {
+			go func(ip string, errChan chan error) {
+				defer s.wg.Done()
+				command := fmt.Sprintf("rm -rf %s;", filepath.Join(s.upDataPath[ip], s.clusterOptions.Version))
+				_, err := s.executor.ExecCommand(ip, command)
+				if err != nil {
+					errChan <- err
+				}
+			}(ip, errChan)
+		}
+		s.wg.Wait()
+		close(errChan)
+	}()
+
+	var has_err = false
+	go func() {
+		defer wgp.Done()
+		for {
+			err, ok := <-errChan
+			if !ok {
+				break
+			}
+			fmt.Println(err)
```

Comment:
There is an infinite loop here

Reply:
The loop will break when the channel is closed.